### ISSUE TITLE
`terragrunt apply-all` fails intermittently with many components using similar providers

### ISSUE THREAD

**Issue by Martin-Idel-SI** [gruntwork-io/terragrunt#879]
Using `terragrunt apply-all`, we see intermittent failure of our pipeline with the error:

```
Error: Could not satisfy plugin requirements

Plugin reinitialization required. Please run "terraform init".
Plugins are external binaries that Terraform uses to access and manipulate
resources. The configuration provided requires plugins which can't be located,
don't satisfy the version constraints, or are otherwise incompatible.
Terraform automatically discovers provider requirements from your
configuration, including providers used in child modules. To see the
requirements and constraints from each module, run "terraform providers".

Error: provider.random: new or changed plugin executable
```

We can reproduce this issue locally and from what I can see it seems to be related to the fact that terragrunt will start all tasks in parallel which will download the plugin to the global cache multiple times as well as maybe access the plugin before it is fully downloaded/copied to the global cache.

---

Here is how I could reproduce the issue locally:

- System: Ubuntu Linux 18.04 (as VM on a Windows host)
- Using global terraform cache with `plugin_cache_dir = "$HOME/.terraform.d/plugin-cache"`

As dummy project, see the attached`test.tar.gz` which contains the following structure:
```
├── module
│   ├── plugin1
│   │   └── main.tf
│   ├── plugin2
│   │   └── main.tf
│   └── plugin3
│       └── main.tf
└── system
    ├── dependencies
    │   └── terragrunt.hcl
    ├── system1
    │   └── terragrunt.hcl
    ├── system2
    │   └── terragrunt.hcl
    ├── system3
    │   └── terragrunt.hcl
    ├── system4
    │   └── terragrunt.hcl
    └── system5
        └── terragrunt.hcl
```
All components are based on some of the "plugins", the "plugins" only use the random and/or external providers. Some of the components depend on each other.

When running `terragrunt apply-all` in the `system` subfolder after removing all `.terragrunt_cache`s and the global plugin cache before running again, I saw the following cases:
- Everything works and the command exits successfully
- Some systems fail with the error above and looking at the `.terragrunt_cache` directories and the `lock.json` files, one can see that the check-sums of some of the providers are wrong in the `lock.json` (see attached `logwronglock.txt`. At the end of the log, I attached the `lock.json` content of various systems showing a mismatch between providers)
- Some systems fail with the error above but the `lock.json` as well as the actual check-sum of the providers in the `.terragrunt_cache` directories seems fine when checking them manually (see attached `logwithoutlockerrors.txt`. At the end of the log, I attached the `lock.json` content of various systems. The checksum of all plugins was identical to the one in the respective `lock.json`).

Logs done using `TF_LOG=trace` show that the providers are actually downloaded multiple times.

[test.tar.gz](https://github.com/gruntwork-io/terragrunt/files/3652741/test.tar.gz)
[logwronglock.txt](https://github.com/gruntwork-io/terragrunt/files/3652744/logwronglock.txt)
[logwithoutlockerrors.txt](https://github.com/gruntwork-io/terragrunt/files/3652747/logwithoutlockerrors.txt)

---

A possible workaround which seems to work locally is to introduce a new system "dep" which downloads all providers and is a direct dependency of all systems. That way, the system "dep" will be initialized first before all other systems and the plugin is already in the global cache when the other systems are applied/initialized in parallel.

**Comment by yorinasub17** [comment]
We've also observed recently that `terraform` starting with tf12 has struggled when you have many instances of it running in parallel on a single box. We've observed this in our usage of terratest in our modules as well, regardless of the global plugin cache.

Note that we can't really do much in `apply-all` to prevent this behavior because in most cases, you will not have a global plugin cache and thus each provider needs to be initialized in each module. We also don't want to start introducing features that parse the terraform code (which is what would be required to know what plugins need to be installed, and how many times), since that goes against the philosophy of terragrunt.  I think the `dep` module approach you have listed is actually a viable workaround for this.

An alternative solution here would be to wait for https://github.com/gruntwork-io/terragrunt/issues/698, which has an open PR in https://github.com/gruntwork-io/terragrunt/pull/636. This will allow you to control the parallelism of terragrunt in the `xxx-all` commands.

Another short term solution we could try would be to [enhance auto retry](https://github.com/gruntwork-io/terragrunt#auto-retry) to include these failures in the list.

**Comment by yorinasub17** [comment]
@brikis98 do you have any other potential ideas to mitigate this?

**Comment by sschu** [comment]
Wouldn't it be possible to run the terraform inits strictly sequential?

**Comment by yorinasub17** [comment]
> Wouldn't it be possible to run the terraform inits strictly sequential?

Sure, but that would be slow and this isn't necessarily an easy logic to implement because `init` is automatically run as part of the sub command (in the auto-init detection feature), not at the stack code level. The configurable parallelism feature is actually going to be the solution that practically implements what you are suggesting.

**Comment by sschu** [comment]
If I understand this feature correctly, it only limits the number of processes running in parallel. Which modules can be terraform-applied concurrently is still controlled by module dependencies. So I assume we will have to tell the user to prevent parallelism of modules using the same provider (e.g. create dependencies between all modules using the same provider)?

**Comment by yorinasub17** [comment]
You can use dependencies to control which modules are run in parallel, but with this feature, you can set an upper bound on how many run concurrently in parallel. So, for example, you can have all the modules apply synchronously using parallelism of 1, regardless of how your dependencies are set up (with the dependencies only controlling the order of execution).

Another thought: to simulate a `init` synchronously followed by apply parallely, you can run `validate-all` with a parallelism of 1 (this is not the same as `init-all` which doesn't exist, but is close enough because the validation step should be reasonably fast) and then `apply-all` with a higher parallelism.

**Comment by sschu** [comment]
Thanks, sounds like a plan. We'll give it a try.

**Comment by ChristophShyper** [comment]
Hi. I just froze my Mac Pro running `validate-all` and found this thread looking for a solution.

@sschu nad @yorinasub17 keep in mind `-parallelism=n` applies only to Terraform modules. So you only limit threads of Terraform, not Terragrunt. Which will still produce heavy load. And they work only on specific Terraform commands.

My solution for one was to make a list of Terragrunt modules directories that can be run in parallel and iterate Terragrunt *-all commands over this list.

**Comment by yorinasub17** [comment]
Oh I think I may have misunderstood the questions, as I was commenting in the context of:

> An alternative solution here would be to wait for #698, which has an open PR in #636. This will allow you to control the parallelism of terragrunt in the xxx-all commands.

**Comment by levkohimins** [comment]
Hi @Martin-Idel-SI, 

The issue should already be resolved in [v0.50.11](https://github.com/gruntwork-io/terragrunt/releases/tag/v0.50.11) release. Therefore I am closing the issue.
